### PR TITLE
fix: Correct modifier key handling in kitty protocol

### PR DESCRIPTION
**Pull request by gabyx**
**What does this PR do:**
Fix a bug in handling modifiers correctly for the kitty protocol.

Fixes: #4247, #4436, #4589

**How was it done?:**
There is a problem in `encode_kitty` function which I tried to debug and hopefully fixed it.
I guess the boolean `use_legacy` which determines to [encode legacy text keys](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#legacy-text-keys) is wrong. 

- In the original any modifiers as `SUPER` (in nvim which sets `KittyKeyboardFlags` to `1`)
  went through this legacy `if` which is wrong. 
- The legacy text keys handling also forgot to add the `Shift` case which I added. [Point 3 here](https://sw.kovidgoyal.net/kitty/keyboard-protocol/#legacy-text-keys).

- I added `Meta` and `Hyper` which I try to add in a further PR)

@wez : As the whole protocol handling is quite difficult and I could not match quite all switched to the protocol.

### CODE REVIEW

File: wezterm-input-types/src/lib.rs [hunk 1]
```diff
@@ -1808,31 +1820,32 @@ impl KeyEvent {
...
                     let mut output = String::new();
                     if self.modifiers.contains(Modifiers::ALT) {
                         output.push('\x1b');
                     }
+
                     if self.modifiers.contains(Modifiers::CTRL) {
                         csi_u_encode(
                             &mut output,
-                            shifted_key.to_ascii_uppercase(),
+                            shifted_key.to_ascii_uppercase(), // Is this to uppercase intended?
```
Comment (gabyx) [inline wezterm-input-types/src/lib.rs:1840]: @wez : Is this intended to turn it into uppercase here? Or is this a typo/mistake where one wanted to tackle point 3 in the protocol : https://sw.kovidgoyal.net/kitty/keyboard-protocol/#legacy-text-keys

File: wezterm-input-types/src/lib.rs [hunk 2]
```diff
@@ -516,6 +518,10 @@ impl TryFrom<String> for Modifiers {
                 mods |= Modifiers::CTRL;
             } else if ele == "SUPER" || ele == "CMD" || ele == "WIN" {
                 mods |= Modifiers::SUPER;
+            } else if ele == "META" {
```
Comment (wez) [inline wezterm-input-types/src/lib.rs:521]: Defining `META` as its own thing needs some care and attention throughout the wezterm crates as well as the termwiz crate, as we've always treated it as an alias of `ALT`, which you can see happening on line 515 above.

This particular line in this PR has no effect because it is evaluated after that line.

Breaking it out into its own thing needs to consider:
* What happens for non-kitty encodings when this modifier is used? Is it simply ignored?
* How does it get set? This PR doesn't seem to add any OS-specific logic to map the incoming key press from the windowing layer to actually set either the META or HYPER flags
* What happens to folks that have defined key assignments using META as the modifier? Do those effectively stop working when they upgrade? Is that ok? (probably?)

File: wezterm-input-types/src/lib.rs [hunk 3]
```diff
@@ -1800,30 +1812,31 @@ impl KeyEvent {
...
                     && is_legacy_key
                     && !(flags.contains(KittyKeyboardFlags::DISAMBIGUATE_ESCAPE_CODES)
                         && (self.modifiers.contains(Modifiers::CTRL)
                             || self.modifiers.contains(Modifiers::ALT)))
-                {
-                    use_legacy = true;
-                }
+                    && !self.modifiers.contains(Modifiers::SUPER)
+                    && !self.modifiers.contains(Modifiers::META)
+                    && !self.modifiers.contains(Modifiers::HYPER);
```
Comment (wez) [inline wezterm-input-types/src/lib.rs:1830]: This is probably a bit more elegant if rephrased as something like:

```rust
!self.modifiers.intersects(Modifiers::SUPER|Modifiers::META|Modifiers::HYPER)
```

File: wezterm-input-types/src/lib.rs [hunk 4]
```diff
@@ -3054,4 +3068,54 @@ mod test {
...
+                modifiers: Modifiers::META,
+                leds: KeyboardLedStatus::empty(),
+                repeat_count: 1,
+                key_is_down: true,
+                raw: None,
+                #[cfg(windows)]
+                win32_uni_char: None,
+            }
+            .encode_kitty(flags),
+            "q".to_string()
```
Comment (wez) [inline wezterm-input-types/src/lib.rs:3118]: I think these test expectations are incorrect; these tests are failing

**Review by wez — COMMENTED** [review]
Thanks for this! There's some more stuff to consider than is just in this PR.

Something that is absent is logic to populate these new key flags.
The only platform that supports those is unix.  IIRC, this is the function that is responsible:

https://github.com/wez/wezterm/blob/fde926722fb6ef05fb3be78624aff33095a283d1/window/src/os/x11/keyboard.rs#L383

File: wezterm-input-types/src/lib.rs (hunk 3 above)
Comment (gabyx) [reply wezterm-input-types/src/lib.rs:1830]: Ah jeah, I was looking for that thing, something like `modifiers.any` but didnt exist and was too dump to construct this. Much better!

File: wezterm-input-types/src/lib.rs (hunk 4 above)
Comment (gabyx) [reply wezterm-input-types/src/lib.rs:3118]: Jup still need to update these. I will reiterate till the test work, no worry.

**Comment by gabyx** [comment]
@wez : Thanks for the reply, I  am not yet done fixing my tests, did not yet adapt the asserts (will do!).

- Thanks for the hint to populate the flags. So that means Meta and Hyper will not work on MacOS?.
- Is it possible to leave the modifiers in this PR and come up with another one to gracefullt add these under Unix at least?
  
 I am still not quite sure, if the kitty thing is all correct, but at least it seams so for the legacy_key part.
 I truely appreciate what you did here, did you look at the kitty's source code for all that `encode_kitty` stuff?
 Its such a convoluted complicated spec haha.

**Comment by wez** [comment]
Only X11 (and Wayland) have META and HYPER modifiers as a concept. Neither Windows nor macOS have equivalents.

I think it is reasonable to define these new modifier flags. I don't think we need to go to extreme lengths to handle META/HYPER elsewhere or for backwards compatibility, but we do need to consider how to handle them for the non-kitty keyboard scenario, which may just be "pretend they don't exist".  Will need to compare and contrast behavior with other unix terminal emulators when using those keys.

I did not read the kitty source for this; only the spec docs, which I agree are not especially clear in a couple of places.

**Comment by gabyx** [comment]
@wez: Ok, thanks for the information about Meta and Hyper. This is very useful. So to not clutter up this PR as a bugfix. lets go with the pragmatic approach and first merge this fix. Like it is presented now.
Is that good enough? Or anything else needed?

- So basically this PR fixes `SUPER` modifier key handling.

**Comment by wez** [comment]
I think this PR is mergeable as-is.

Re: other modifiers and the future, are you proposing the addition of `MOD1..MOD5` to the `Modifiers` type in `wezterm-input-types`?  I think that those could be added (even though eg: `ALT` is already implicitly handled under a different name) in a similar way to the `LEFT_XXX` and `RIGHT_XXX` flags; those are populated if the system can make the distinction, but are otherwise non-essential and can then be inspected by eg: the kitty protocol encoding logic.

A potential wrinkle with that is: the flags type is 16 bits wide and there aren't enough available bits to represent 5 additional modifiers, so perhaps we should just add the mods that are not currently represented.  I'd like to avoid making `Modifiers` 32 bits wide if we can avoid it, but if it doesn't change the size of aggregate types (like the size of the entries in various key tables, where the cardinality * the size of the struct represents an increase in RAM usage) then there's no real argument against it.

**Comment by gabyx** [comment]
@wez: Jeah it might be the only sane way to add Mod1-Mod5 to the table an check against where the other xkb types are handled. This addition will not hurt wezterm. But we still dont know which keys Mod1-Mod5 will be what “thing” for the kitty protocol. (to what Mod? did Hyper map?)

I think, Kitty is doing this by looking at the code: its checking the virtual modifier table with x11 calls and xkb and this is somewhat involved and as far as I could grasp was that this will correlate Super_L and other modifiers like Hyper_L/R to the 9 different modifiers xkb knows (Shift, Ctrl, Alt, Lock, Mod1-5) (`xmodmap -pm`). When we know this table (kitty does that at startup) 
we know for lets say Hyper_L maps to Mod3 (the user chose so) , if Mod3 was pressed, then we can encode a “Hyper” modifier in the kitty protocol.

So basically function: https://github.com/wez/wezterm/blob/fde926722fb6ef05fb3be78624aff33095a283d1/window/src/os/x11/keyboard.rs#L383

is somewhat only partially correct, it checks for Mod4 and relating this to Super. But the user might have changed this, it maybe has chosen that a certain key activates Super_L an mapped this to Mod3.

I hope my explanations were correct and understandable.
I think without implementing the exact same code as in kitty (will link to it, later)
wezterm preassumes some standard mapping from Alt, Shift, Ctrl, Super to Mod1-5.

**Comment by gabyx** [comment]
The other thing is I am unaware of if Wayland has some otherstuff, it still uses xkb afaik. And the virtual modifier table is still a thing. Maybe we could really port the C code from kitty into rust: It basically this function: https://github.com/kovidgoyal/kitty/blob/72e14053f55b0165a47d6cd743a0c1de1944149a/glfw/xkb_glfw.c#L376

**Comment by wez** [comment]
Thanks!
